packages/dataviews/src/dataform.tsx

```diff
@@ -0,0 +1,94 @@
+/**
+ * External dependencies
+ */
+import type { FormEventHandler } from 'react'; // TODO: how to import FormEventHandler?
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import {
+	Button,
+	TextControl,
+	__experimentalHStack as HStack,
+	__experimentalVStack as VStack,
+} from '@wordpress/components';
+import { useCallback, useMemo, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import type { NormalizedField } from './types';
+
+type DataFormProps< Item > = {
+	item: Item;
+	onUpdateItem: ( item: Item ) => void;
+	fields: NormalizedField< Item >[]; // TODO: use Field. Normalize them first.
+	form: {
+		closeForm: () => void;
+		isBusy: boolean;
+		onSubmitLabel?: string;
+		onCancelLabel?: string;
+	};
+};
+
+export default function DataForm< Item >( {
+	item: data,
+	onUpdateItem,
+	fields,
+	form: { closeForm, isBusy, onCancelLabel, onSubmitLabel },
+}: DataFormProps< Item > ) {
+	const [ item, setItem ] = useState( data );
+	const onSubmit: FormEventHandler< HTMLFormElement > = useCallback(
+		( event ) => {
+			event.preventDefault();
+
+			onUpdateItem( item );
+		},
+		[ item ]
+	);
+
+	// TODO: make it work for all fields.
+	const onChange = useCallback( ( title: string ) => {
+		setItem( { ...item, title } );
+	}, [] );
+
+	const components = useMemo(
+		() =>
+			fields.map( ( field ) => {
+				if ( field.type === 'text' ) {
```

Comment:
Alternatively, instead of providing a `type` the field could offer an `edit` view to provide its own control. I found this to be an advance use case, so I decided to start with the `type`.

This is also a reason why starting with the actions is a good way to scope down the PR. The form rendered here needs to provide an input component, which we don't have in the Fields API yet — the `render` is only used in DataViews for now. When we get there, we need to consider whether the `render` can be shared between `DataViews` and `DataForms`. We already have situations where each type of view in DataViews needs different things, so it's not unthinkable that DataForms would need its own. I thought it'd be better to explore that problem space in a separate PR.